### PR TITLE
JNG-4577 transfer object constructor

### PR DESCRIPTION
**Pull request by robertcsakany**
<table>
<td>
  <a href="https://blackbelt.atlassian.net/browse/JNG-4577" title="JNG-4577" target="_blank">StoryJNG-4577</a>  Transform JSL Service layer to PSM
  </td></table>
  <br />

### CODE REVIEW

**Review by Madmat8 — APPROVED** [review]

**Review by noherczeg — APPROVED** [review]

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/data/primitiveTypedElement.etl [hunk 1]
```diff
@@ -152,3 +152,66 @@ rule CreateReadsDataPropertyForMappedTransferObjectTranferFieldDeclaration
...
+
+        t.setId("(jsl/" + s.field.reference.getId() + ")/CreateDefaultStaticDataForUnmappedTransferObjectConstructor");
+        t.name = defaultDefaultNamePrefix + s.field.reference.name + defaultDefaultNameMidfix + s.field.reference.eContainer.name + defaultDefaultNamePostfix;
+        t.dataType = s.field.reference.referenceType.getPSMEquivalent();
+        t.required = false;
+
+        s.eContainer.eContainer.eContainer.getModelRoot().elements.add(t);
+        log.debug("Created StaticData for Transfer Object Default Value: " + s);
+    }
+
```
Comment (noherczeg) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/data/primitiveTypedElement.etl]: Hah!

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferAttribute.etl [hunk 2]
```diff
@@ -57,8 +63,62 @@ rule CreateMappedTransferAttribute
...
+
+@greedy
+rule CreateTransferEntityDefaultValueAttributeForUnmappedTransferObjectConstructor
+    transform s: JSL!TransferDefault
+    to t: JUDOPSM!TransferAttribute {
+        guard: s.field.reference.referenceType.isKindOf(JSL!PrimitiveDeclaration) and s.field.reference.eContainer.map.isUndefined()
+        t.name = defaultDefaultNamePrefix + s.field.reference.name + defaultDefaultNameMidfix + s.field.reference.eContainer.name + defaultDefaultNamePostfix;
+        t.binding = s.equivalent("CreateDefaultStaticDataForUnmappedTransferObjectConstructor");	  
+        t.dataType = s.field.reference.referenceType.getPSMEquivalent();
+        t.setId("(jsl/" + s.field.reference.getId() + ")/CreateTransferEntityDefaultValueRelationForUnmappedTransferObjectConstructor");
```
Comment (bencegelei) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferAttribute.etl]: > CreateTransferEntityDefaultValue**Relation**ForUnmappedTransferObjectConstructor

Should be CreateTransferEntityDefaultValue**Attribute**ForUnmappedTransferObjectConstructor

File: judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestCreateMappedTransferObjectTypeModel.jsl [hunk 3]
```diff
@@ -37,16 +37,15 @@ transfer Mapped maps Entity as e {
 	field Integer mappedAttribute maps e.attribute;
 	field MappedRelated mappedContainment maps e.containment;
 	field MappedRelated[] mappedContainmentCollection maps e.containmentCollection;
-    // TODO: JNG-4603 Fix associations
-    // field MappedRelated mappedAssociation maps e.association;
-    // field MappedRelated mappedAssociationOpposite maps e.entityRelatedOpposite;
+
+    field MappedRelated mappedAssociation maps e.association;
```
Comment (gaborflorian) [inline judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestCreateMappedTransferObjectTypeModel.jsl:41]: pls add a tab

File: judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestCreateMappedTransferObjectTypeModel.jsl [hunk 4]
```diff
@@ -37,16 +37,15 @@ transfer Mapped maps Entity as e {
...
+    field MappedRelated mappedAssociationOpposite maps e.entityRelatedOpposite;
 
 	field Integer derivedAttribute reads e.attribute;
 	field MappedRelated derivedContainment reads e.containment;
 	field MappedRelated[] derivedContainmentCollection reads e.containmentCollection;    
     field MappedRelated derivedAssociation reads e.association;
-    // TODO: JNG-4603 Fix associations
-    // field MappedRelated derivedAssociationOpposite maps e.entityRelatedOpposite;
+    field MappedRelated derivedAssociationOpposite maps e.entityRelatedOpposite;
```
Comment (gaborflorian) [inline judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestCreateMappedTransferObjectTypeModel.jsl:48]: pls add here tabs too

File: judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestTransferObjectConstructorModel.jsl [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+		self.mappedAssociationOpposite = EntityRelated!any();
+		self.mappedAssociationOppositeCollection = EntityRelated!all();
+		self.transientRelation = EntityRelated!any();
+		self.transientRelationCollection = EntityRelated!all();
+	}
+}
+
+transfer MappedRelated maps EntityRelated as e {
+	field Integer mappedAttribute maps e.attribute;
+}
```
Comment (bencegelei) [inline judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestTransferObjectConstructorModel.jsl]: newline EOF

**Review by bencegelei — CHANGES_REQUESTED** [review]

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/data/primitiveTypedElement.etl (hunk 1 above)
Comment (bencegelei) [reply judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/data/primitiveTypedElement.etl]: newline EOF (github does not mark missing newline EOF if not in "Files" view)

File: judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestCreateMappedTransferObjectTypeModel.jsl (hunk 4 above)
Comment (robertcsakany) [reply judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestCreateMappedTransferObjectTypeModel.jsl:48]: Its your replace will fix this issue

File: judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestCreateMappedTransferObjectTypeModel.jsl (hunk 3 above)
Comment (robertcsakany) [reply judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestCreateMappedTransferObjectTypeModel.jsl:41]: Its your replace will fix this issue

File: judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestTransferObjectConstructorModel.jsl (hunk 5 above)
Comment (robertcsakany) [reply judo-tatami-jsl-jsl2psm/src/test/resources/transferobject/TestTransferObjectConstructorModel.jsl]: Gabor's fix will resolve this.

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/data/primitiveTypedElement.etl (hunk 1 above)
Comment (robertcsakany) [reply judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/data/primitiveTypedElement.etl]: Gabor's replace will fox this issue.

**Review by bencegelei — APPROVED** [review]

**Pull request merged**